flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala

```diff
@@ -1395,50 +1399,25 @@ object AggregateUtil {
 
         case udagg: AggSqlFunction =>
           aggregates(index) = udagg.getFunction
+          accTypes(index) = udagg.accType
 
         case unSupported: SqlAggFunction =>
           throw new TableException(s"unsupported Function: '${unSupported.getName}'")
       }
     }
 
-    (aggFieldIndexes, aggregates)
-  }
-
-  private def createAccumulatorType(
-      aggregates: Array[TableAggregateFunction[_, _]]): Seq[TypeInformation[_]] = {
-
-    val aggTypes: Seq[TypeInformation[_]] =
-      aggregates.map {
-        agg =>
-          val accType = try {
-            val method: Method = agg.getClass.getMethod("getAccumulatorType")
-            method.invoke(agg).asInstanceOf[TypeInformation[_]]
-          } catch {
-            case _: NoSuchMethodException => null
-            case ite: Throwable => throw new TableException("Unexpected exception:", ite)
-          }
-          if (accType != null) {
-            accType
-          } else {
-            val accumulator = agg.createAccumulator()
-            try {
-              TypeInformation.of(accumulator.getClass)
-            } catch {
-              case ite: InvalidTypesException =>
-                throw new TableException(
-                  "Cannot infer type of accumulator. " +
-                    "You can override AggregateFunction.getAccumulatorType() to specify the type.",
-                  ite)
-            }
-          }
-      }
+    // create accumulator type information for every aggregate function
+    aggregates.zipWithIndex.foreach { case (agg, index) =>
+      accTypes(index) = getAccumulatorTypeOfAggregateFunction(agg, accTypes(index))
```

Comment:
Yes, adding the `null == accTypes(index)` condition can reduce one more time calling `getAccumulatorTypeOfAggregateFunction`.  I will add it.